### PR TITLE
[mono][aot] Enabling AOT wrappers for virtual delegates

### PR DESCRIPTION
**Pull request by LeVladIonescu**
Adding AOT wrappers for delegate invoke virtual.
Contributes to https://github.com/dotnet/runtime/issues/83329.

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -3778,6 +3778,8 @@ encode_method_ref (MonoAotCompile *acfg, MonoMethod *method, guint8 *buf, guint8
 	MonoJumpInfoToken *ji;
 	guint8 *p = buf;
 
+	printf("NAME = %s\n", method->name);
```
Comment (LeVladIonescu) [inline src/mono/mono/mini/aot-compiler.c:3781]: unrelated change...only for debug reasons

**Comment by LeVladIonescu** [comment]
Ready to review / merge.
Library tests build failures from CI are related.

**Review by vargaz — APPROVED** [review]

**Comment by SamMonoRT** [comment]
@LeVladIonescu - as part of this PR consider enabling any disabled tests fixed by the changes.

**Comment by LeVladIonescu** [comment]
@SamMonoRT will check

File: src/mono/mono/metadata/metadata.c [hunk 2]
```diff
@@ -2512,6 +2512,33 @@ mono_metadata_signature_dup (MonoMethodSignature *sig)
...
+MonoMethodSignature*
+mono_metadata_signature_dup_delegate_invoke_to_target (MonoMethodSignature *sig)
+{
+	MonoMethodSignature *res = mono_metadata_signature_dup_full (NULL, sig);
+
+	for (int i = 0 ; i < sig->param_count - 1; i ++) {
+		res->params [i] = sig->params [i + 1];
+	}
+
+    res->param_count --;
```
Comment (kotlarmilos) [inline src/mono/mono/metadata/metadata.c:2534]: Nit: Fix formatting and remove commented code

File: src/mono/mono/mini/aot-compiler.c [hunk 3]
```diff
@@ -4546,6 +4550,39 @@ can_marshal_struct (MonoClass *klass)
...
+	MonoGenericInst *inst;
+	MonoGenericContainer *container;
+
+	memset (ctx, 0, sizeof (MonoGenericContext));
+
+	if (mono_class_is_gtd (method->klass)) {
+		shared_context = mono_class_get_generic_container (method->klass)->context;
+		inst = shared_context.class_inst;
+
+		args = g_new0 (MonoType*, inst->type_argc);
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:4568]: Is the memory being deallocated correctly?

Comment (LeVladIonescu) [reply src/mono/mono/mini/aot-compiler.c:4568]: No, it is not. I would suggest deallocating it after https://github.com/dotnet/runtime/blob/175dae98033a8a03627fbcdaa8575ad3a3fca512/src/mono/mono/metadata/metadata.c#L3399

Comment (vargaz) [reply src/mono/mono/mini/aot-compiler.c:4568]: I can be freed, but the aot compiler is already leaking memory in a number of places, so its not that important.

**Comment by LeVladIonescu** [comment]
Current failures are known and seem unrelated:
- https://github.com/dotnet/runtime/issues/82637
- https://github.com/dotnet/runtime/issues/80619
